service/storage_service.cc

```diff
@@ -7493,20 +7495,59 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
             throw std::runtime_error(format("Cannot stream within the same node, tablet: {}, shard {} -> {}",
                                             tablet, leaving_replica.shard, trinfo->pending_replica.shard));
         }
```

Comment:
```
        locator::tablet_replica leaving_replica = locator::get_leaving_replica(tinfo, *trinfo);
        if (leaving_replica.host == tm->get_my_id()) {
            // The algorithm doesn't work with tablet migration within the same node because
            // it assumes there is only one tablet replica, picked by the sharder, on local node.
            throw std::runtime_error(format("Cannot stream within the same node, tablet: {}, shard {} -> {}",
                                            tablet, leaving_replica.shard, trinfo->pending_replica.shard));
        }
```
This check should be in the locator::tablet_transition_kind::migration block.
The leaving node in replace and removenode case will not be the target of the
storage_service::stream_tablet rpc at all.

Reply:
With regular migration,  the node receives stream_tablet rpc IS the target of migration, no?  It pulls data from other replica node. What's why we check leaving_replica.host (the src node) should not be the node itself (the dst node which pulls data from the src node), which checks src of migration != dst of migration. 

With replace and removenode, the node receives  stream_tablet rpc is the target of migration (rebuild migration kind, which pulls data from existing replicas). The leaving_replica is the dead node (to be replaced or removed). 
There is no point to check leaving_replica != the node which pulls the data and builds the replica because the code below never streams data from the leaving_replica at all. If we really want to check stream src != dst node, we should check the `host2ip(r.host)` below != the node itself, not leaving_replica for the locator::tablet_transition_kind::rebuild case. 

```
                for (auto r : trinfo->next) {
                    if (r.host != trinfo->pending_replica.host && !excluded.contains(raft::server_id(r.host.uuid()))) {
                        ranges_per_endpoint[host2ip(r.host)].emplace_back(range);
                    }
                }
```

Reply:
I see what you mean. But the check for leaving.host != pending.host is not guarding streaming, it's guarding replication change. We don't support change in replication which involves the same node (yet). That's intra-node tablet migration. So the invariant should hold for all tablet transitions, regardless of streaming. We could stream nothing and it should still hold.